File: src/main/scala/VectorFactory.scala
object VectorFactory {
  private var types: Map[String, VectorType] = Map(
    "Polar Vector" -> new PolarVectorType,
    "Cartesian Vector" -> new CartesianVectorType
  )

  def registerType(name: String, vectorType: VectorType): Unit = {
    types += (name -> vectorType)
  }

  def createVector(typeName: String): Vector = {
    types(typeName).create
  }
}

File: src/main/kotlin/VectorFactory.kt
object VectorFactory {
    private var types: MutableMap<String, VectorType> = mutableMapOf(
        "Polar Vector" to PolarVectorType(),
        "Cartesian Vector" to CartesianVectorType()
    )

    fun registerType(name: String, vectorType: VectorType) {
        types[name] = vectorType
    }

    fun createVector(typeName: String): Vector {
        return types[typeName]?.create() ?: throw IllegalArgumentException("No such vector type registered")
    }
}
